### ISSUE TITLE
N+ requests when use include feature 

### ISSUE THREAD

**Issue by youzik** [Netflix/fast_jsonapi#246]
Hi there. 

My frontend team helped me to find the fun issue. Their deserialization library requires included models for deserializing `has_many` associations. That's not a problem, but when I include models fast_jsonapi makes queries to fetch ids for `has_many` fields and more queries for included fields. 

For example
Without include.
```
(10.1ms)  SELECT * FROM doctors <...>
(4.0ms)  SELECT "clinics"."id" FROM "clinics" <...>
(4.0ms)  SELECT "clinics"."id" FROM "clinics" <...>
```

Trying to include
```
(7.2ms)  SELECT * FROM doctors <...>
(4.1ms)  SELECT "clinics"."id" FROM "clinics" <...>
(3.8ms)  SELECT "clinics"."id" FROM "clinics" <...>
Practice Load (4.8ms)  SELECT "clinics".* FROM "clinics" <...>
Office Load (3.5ms)  SELECT "clinics".* FROM "clinics" <...>
```

Just wondered if we see included models we can load this model to memory and use id fields for has_many with no additional queries?

Should i work on this feature and open a pull request?

**Comment by shishirmk** [comment]
We are trying to remain agnostic to which underlying datastore and library you are using. So we dont think it makes sense to add this logic to the serializer. However we avoid getting into N+1 queries by using the ActiveRecord includes feature https://apidock.com/rails/ActiveRecord/QueryMethods/includes.

**Comment by ababich** [comment]
I do not see N+1 issue with `fast_jsonapi` in all cases except `has_many` relation when return a group of "ones" with lots of "manys" included

**Comment by youzik** [comment]
@ababich that's case I faced during my research

**Comment by christophersansone** [comment]
I took a quick look to see what could be possible here, but I have to agree with @shishirmk.

It's the job of `fast_jsonapi` to serialize the data given to it as fast as possible.  It's not the job of `fast_jsonapi` to ensure that the data sent to it is optimized for performance.  The data should be optimized before it is passed into the serializer.

It looks to me like you can solve by changing your relation from `Doctor.all` to `Doctor.includes(:clinics)`.

Maybe we can add some additional performance tips in the `README` for things like this.

**Comment by ababich** [comment]
@shishirmk / @christophersansone I think you need to provide examples how to do so at least here before updating `README`

**Comment by christophersansone** [comment]
@ababich How about this:

```ruby
class MoviesController < ApplicationController
  def index
    # select the records
    records = Movie.where('year > 2010')
    # include the relations ahead of time to avoid n+1 queries
    relation_to_serialize = movies.includes(:actors)
    # instantiate the serializer
    serializer = MovieSerializer.new(relation_to_serialize)
    # perform the serialization
    render json: serializer.serializable_hash
  end
end
```

**Comment by youzik** [comment]
@christophersansone i guess you didn't catch my idea. I mean it's a little weird when gem asks for the same data twice per request. Once for render 'relationship' block, and then for 'includes' block. Why we just can't store ids from `select * from <>` query and use it later for 'relationship'?

**Comment by christophersansone** [comment]
@youzik Sorry, you're right, I guess I'm not following you.

I am assuming you are using `ActiveRecord`, correct?

Let's make a real world example here.  Assume this model:

```ruby
class Movie < ActiveRecord::Base
  has_many :actors
end
```

And this serializer:

```ruby
class MovieSerializer
  include FastJsonapi::ObjectSerializer
  has_many :actors
end
```

And finally calling the serializer:

```ruby
movies = Movie.where('year > 2010').includes(:actors)
json = MovieSerializer.new(movies, { include: [:actors] }).serialized_json
```

When the serializer performs the serialization on a `movie` record, it will simply call `movie.actors` to get the list of actors.  And yes, it will call `movie.actors` once for the `relationships` and then a second time for the `include`.  This gem relies on the underlying model to optimize data fetching and memoize the data.  In the case of ActiveRecord, the first call to `actors` will execute the query, and the results are indeed cached at the ActiveRecord level.  If you use `.includes(:actors)` then it will only make one database request, but if you omit that, it will make n+1 queries.  Either way, the data is indeed cached.  At this point, it seems the consensus of the maintainers is that it's not the job of this gem to be the cache, and in the case of ActiveRecord, AFAIK it is unnecessary and would duplicate the work that ActiveRecord is already doing.

This is my understanding of how it works upon a quick read through the code.  I didn't write it so I may be mistaken, but that's how I would expect it to behave, and the code seems to indicate such.

Hope this helps.  Let me know if you believe this explanation about this gem or ActiveRecord's behavior is incorrect.  Or if I still have not understood the problem at hand, can you please provide some complete examples?  Perhaps even put together a sample application and then include the log?

**Comment by ababich** [comment]
I think you rely on AR too much. I will prepare Sequel use cases for the issue

**Comment by christophersansone** [comment]
@ababich Thanks, looking forward to seeing how you use this gem with Sequel.

Can we get confirmation, perhaps by @youzik, that ActiveRecord + this gem is working as expected with eager loading and caching?

**Comment by youzik** [comment]
@christophersansone yeah, you're right here! We used to handle this issue same way, and allow the serializer to use cached data as many times as it wants. Anyway, in our scenario, it's not the best way, for some architecture reasons, but I assume we software developers, not just framework tuners, so we faced issues every day :). Thanks for your help and your time. I'll be happy to see how @ababich handle it with Sequel too.
